Реализован функционал ТЗ №9

OlegDyuzhin opened the pull request.

OlegDyuzhin committed:
Реализован функционал ТЗ №9

avfyodorov commented on src/server/DurationAdapter.java:
```diff
@@ -0,0 +1,31 @@
+import java.io.IOException;
+import java.time.Duration;
+
+public class DurationAdapter extends TypeAdapter<Duration> {
```
Самодельные адаптеры лучше выделять в отдельный пакет, например, adapters

OlegDyuzhin replied:
Перенес

avfyodorov commented on src/server/BaseHttpHandler.java:
```diff
@@ -0,0 +1,203 @@
+        }
+    }
+
+    private void readResource(HttpExchange exchange, String[] path) throws IOException {
```
Сейчас получается,  что вся бизнес-логика реализована в этом классе, что не совсем соответствует ООП. Хорошо было бы провести рефакторинг, в базовом классе оставить обработчик и методы-заглушки, а в классах наследниках переопределять только эти методы (processGet, processPost….), если они нужны. Такое решение лучше соответствует принципам OOP 

На Ваше усмотрение. 

```java
       public void handle(HttpExchange exchange) throws IOException { 
       String method = exchange.getRequestMethod();
           switch (method) {
               case "GET":
                   processGet(exchange);
                   break;
               case "POST":
                   processPost(exchange);
                   break;
               case "DELETE":
                   processDelete(exchange);
                   break;
               default:
                   writeToUser(exchange, "Данный метод не предусмотрен");
           }
           
    protected void processGet(....) {
        sendMethodNotAllowed(...);
   }
   protected void processPost(....) {
        sendMethodNotAllowed(...);
   }
```

OlegDyuzhin replied:
Разбил функционал по классам наследникам.
Наверное если делать по context можно былобы сделать абстрактный основной класс, и запускать сервер от контекста таким образом его настраивая и тогда sendMethodNotAllowed(...) работал бы если сервер на этот контекст не настроен.

Спасибо за ревью!

avfyodorov replied:
Да, конечно. Идея как раз была в том, чтобы сделать наследников по типам задач  -
TaskHandler, EpicHandler, HistoryHandler…. А в них реализовывать только те обработчики запросов (GET, POST, DELETE), которые нужны для данного типа. 

```java
public class HistoryHandler extends BaseHandler {
............
    @Override
    protected void processGet(HttpExchange exchange, String path) throws IOException {
.........
```

avfyodorov reviewed (commented):
Добрый вечер, Олег! 

Очень хорошая работа, от меня всего пара небольших уточнений.

OlegDyuzhin committed:
Устранены замечания по ТЗ №9:
-BaseHttpHandler разделен на классы по выполняемым операциям
-Адаптеры перенесены в отдельный пакет
Небольшой рефакторинг:
-В менеджере добавлен генератор Gson со стандартными настройками

avfyodorov reviewed (commented):
Добрый день, Олег!

Замечаний нет.
Работа принята.
